source/Game.cpp

```diff
@@ -0,0 +1,199 @@
+#include"Game.hpp"
+
+#include<sstream>
+#include<iostream>
+#include<cctype>
+#include<string>
+#include"Error.hpp"
+
+namespace dungeon {
+
+Game::Game(): isEnd(false) {
+  initialize();
+}
+
+void Game::welcome() {
+  ios.clear();
+  ios.print("hello, world");
+  ios.pause();
+}
+
+void Game::selectClass() {
+  ios.clear();
+  ios.print("Select your Class");
+  ios.print("A. Saber");
+  ios.print("B. Archer");
+  ios.print("C. Lancer");
+  // ios.print("D. Rider");
+  // ios.print("E. Caster");
+  // ios.print("F. Assassin");
+  ios.print("G. Berserker");
+  switch (ios.inputChar()) {
+  case 'A':
+    mainChar.type = Class::saber; break;
+  case 'B':
+    mainChar.type = Class::archer; break;
+  case 'C':
+    mainChar.type = Class::lancer; break;
+  // case 'D':
+  //   mainChar.type = Class::rider; break;
+  // case 'E':
+  //   mainChar.type = Class::caster; break;
+  // case 'F':
+  //   mainChar.type = Class::assassin; break;
+  case 'G':
+    mainChar.type = Class::berserker; break;
+  default:
+    throw TypeError();
+  }
+}
+
+void Game::enterName() {
+  using namespace std::literals;
+  auto input = ios.input("Enter your name: ");
+  mainChar.name = input.front();
+  for (int i = 1; i < (int)input.size(); i++)
+    mainChar.name += " "s + input[i];
+}
+
+void Game::initialize() {
+  welcome();
+  selectClass();
+  enterName();
+  mainChar.initialize();
+  mainChar.curPos = map.at(0, 0);
+}
+
+void Game::printMapInfo() {
+  std::stringstream ss;
+  ss << "current position in map: "
+     << "x [" << mainChar.curX << "] "
+     << "y [" << mainChar.curY << "]\n";
+  ss << "Current room type: ";
+  switch (mainChar.curPos->type) {
+  case RoomType::nothing:
+    ss << "nothing"; break;
+  case RoomType::treasure:
+    ss << "treasure"; break;
+  case RoomType::npc:
+    ss << "npc"; break;
+  case RoomType::monster:
+    ss << "monster"; break;
+  default:
+    throw TypeError();
+  }
+  ios.print(ss.str());
+}
+
+void Game::move() {
+  ios.print("Where do you want to go?");
+  ios.print("A. Go up");
+  ios.print("B. Go down");
+  ios.print("C. Go left");
+  ios.print("D. Go right");
+  auto input = ios.input();
+  switch (std::toupper(input[0][0])) {
+  case 'A':
+    mainChar.curX -= 1; break;
+  case 'B':
+    mainChar.curX += 1; break;
+  case 'C':
+    mainChar.curY -= 1; break;
+  case 'D':
+    mainChar.curY += 1; break;
+  default:
+    ios.print("wrong input");
+  }
+  mainChar.curPos = map.at(mainChar.curX, mainChar.curY);
+
+// big problem here
+
+  switch (mainChar.curPos->type) {
```

Comment:
You can, for example, create a abstract function `doAction(Game&)` in `Room` so that you can simply write `mainChar.curPos->doAction(*this)` here. It's much better to internalize specializations into child class's implementation as much as possible.